File: src/main/java/org/jukeboxmc/util/Identifier.java
package org.jukeboxmc.util;

import lombok.Getter;

@Getter
public class Identifier {

    private static final String DEFAULT_NAMESPACE = "minecraft";
    private static final char SEPARATOR = ':';

    private static final Identifier EMPTY = new Identifier("", "", String.valueOf(SEPARATOR));

    private final String namespace;
    private final String name;
    private final String fullName;
    private final int hashCode;

    public Identifier( String namespace, String name, String fullName) {
        this.namespace = namespace;
        this.name = name;
        this.fullName = fullName;
        this.hashCode = fullName.hashCode();
    }

    public static Identifier fromString(String str) {
        if(str.isBlank()) return EMPTY;
        str = str.trim();
        final String[] nameParts = str.indexOf(SEPARATOR) != -1 ? str.split(String.valueOf(SEPARATOR)) : new String[]{str};
        final String namespace = nameParts.length > 1 ? nameParts[0] : DEFAULT_NAMESPACE;
        final String name = nameParts.length > 1 ? nameParts[1] : str;
        final String fullName = namespace + SEPARATOR + name;
        return new Identifier(namespace, name, fullName);
    }

    @Override
    public boolean equals(Object o) {
        return o instanceof Identifier && ((Identifier) o).hashCode == this.hashCode;
    }

    @Override
    public int hashCode() {
        return this.hashCode;
    }

    @Override
    public String toString() {
        return this.fullName;
    }

}

File: JukeboxMC-API/src/main/kotlin/org/jukeboxmc/api/Identifier.kt
package org.jukeboxmc.api

class Identifier private constructor(
    private val namespace: String,
    private val name: String,
    private val fullName: String,
    private val hashCodeValue: Int = fullName.hashCode()
) {

    companion object {
        private const val DEFAULT_NAMESPACE = "minecraft"
        private const val SEPARATOR = ':'

        val EMPTY: Identifier = Identifier("", "", SEPARATOR.toString(), SEPARATOR.toString().hashCode())

        fun fromString(str: String): Identifier {
            if (str.isBlank()) EMPTY
            var strValue = str
            strValue = strValue.trim { it <= ' ' }
            val nameParts = if (strValue.indexOf(SEPARATOR) != -1) strValue.split(SEPARATOR.toString().toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray() else arrayOf(strValue)
            val namespace = if (nameParts.size > 1) nameParts[0] else DEFAULT_NAMESPACE
            val name = if (nameParts.size > 1) nameParts[1] else strValue
            val fullName: String = namespace + SEPARATOR + name
            return Identifier(namespace, name, fullName)
        }
    }

    fun getNamespace(): String {
        return this.namespace
    }

    fun getName(): String {
        return this.name
    }

    fun getFullName(): String {
        return this.fullName
    }

    override fun equals(other: Any?): Boolean {
        return other is Identifier && other.hashCodeValue == this.hashCodeValue
    }

    override fun hashCode(): Int {
        return hashCodeValue
    }

    override fun toString(): String {
        return fullName
    }
}